### PR TITLE
add byte_format decoding support and specs

### PR DESCRIPTION
**Pull request by jreinert**

### CODE REVIEW

**Comment by asterite** [comment]
Cool :-)

We have to decide on the names. Right now it's write_bytes/read_object. Before the initial commit it was named write_object, so maybe it should be write_object/read_object or write_bytes/read_bytes, but I don't know.

**Comment by benoist** [comment]
I see you've created a slice io in the spec file. Maybe you can add that to the std lib. That fixes #825

**Comment by jreinert** [comment]
@asterite i'd be leaning towards write_object/read_object. Especially because i could imagine other objects implementing to_io/from_io as well. Right now I'm trying to refactor websockets a bit so that we have websocket frame classes for the different opcodes. There to_io and from_io would make a lot of sense too.

**Comment by Papierkorb** [comment]
I'm a bit confused tbh. I think that read_object and read_bytes (and their write_ counterparts) are fundamentally different methods, with both having a valid purpose:

read_object reads an object, whose definition I don't know. It may do that by calling the objects #from_io method. However the from method is called, I think expecting an object to implement that method is reasonable.

read_bytes reads stuff into a slice I pass into. Compare this to read_object: There I know what I'm getting back (The object I passed into it), while with read_bytes, I don't know it, all I see is arbitrary data.

But, now from a more real-world perspective, both are really useful as well. Suppose Slice(T) implements to_io and from_io. to_io prepends the slices size when writing, so the data written looks like `[Int32 length][UInt8 bytes ...]`. from_io can use this to quickly read back into the slice. Now, we can conveniently store and load arbitrary, binary data. Using the read_bytes/write_bytes methods, we can read/write a slice *plain*, so the read/written data looks like `[UInt8 bytes ...]`.

TL;DR: Both methods have their purpose.

**Comment by jreinert** [comment]
@Papierkorb the read_bytes you're describing is actually read(slice : Slice(UInt8)) which is defined for all Objects that implement IO.

**Comment by asterite** [comment]
Yes, I now think read_object is a better name for the current read_bytes

**Comment by jreinert** [comment]
@asterite i think you meant write_*

**Comment by jreinert** [comment]
anything else you need from this?

**Comment by asterite** [comment]
@jreinert Just a bit of time to sit down with @waj, review it and probably merge it :-)
